### PR TITLE
Add Redux with React Router, React router redux, Jest, Enzyme

### PR DESCRIPTION
**Pull request by thatguynamedandy**

### CODE REVIEW

**Comment by pmsorhaindo** [comment]
LGTM! What is the plan with regard to async action creators? Are we going to continue with thunk or shall we experiment with something else?

**Comment by thatguynamedandy** [comment]
@pmsorhaindo I'm currently playing with redux-saga on a separate branch.

**Comment by pmsorhaindo** [comment]
Cool, also one other comment, maybe the example actions and reducers could go, as they'll likely be removed every time someone clones the project. But I totally get that they might be useful for people starting work from this project for the first time so I'm happy for it to go in if we feel there is a need for it.

**Comment by thatguynamedandy** [comment]
Ideally I want the example app to do one async and one none async task. Nothing particularly useful but just a way of conveying our opinion on best practice. I think that the first step after cloning probably would be to delete to skeleton actions, but if it ends well they should be nicely isolated and serve as a good reference point. So it's good to be extra critical about the approach now.

File: package.json [hunk 1]
```diff
@@ -20,21 +29,34 @@
...
     "react-dom": "^15.6.1",
     "react-hot-loader": "^3.0.0-beta.7",
+    "react-redux": "^5.0.6",
+    "react-router-redux": "next",
+    "react-router": "^4.2.0",
+    "react-test-renderer": "^15.6.1",
     "react": "^15.6.1",
+    "redux-devtools-extension": "^2.13.2",
+    "redux": "^3.7.2",
+    "regenerator-runtime": "^0.11.0",
```
Comment (HHogg) [inline package.json]: You don't need to install `regenerator-runtime` or `babel-jest` [as we're using yarn](https://github.com/facebook/jest#using-babel)

File: src/index.js [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+  ),
+);
 
 function render(Component) {
   ReactDOM.render(
     <AppContainer>
-      <Component />
+      <Provider store={ store }>
+        <ConnectedRouter history={ history }>
+          <div>
```
Comment (HHogg) [inline src/index.js]: Is this div needed?

File: src/containers/App.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { addProduct, removeProduct } from '../store/products';
+import App from '../components/App';
+
+export class AppContainer extends Component {
```
Comment (HHogg) [inline src/containers/App.js]: This component isn't needed as it's just spreading the props. The connected component could just be the imported App component 

```jsx
import App from '../components/App';
export default connect(mapStateToProps, mapDispatchToProps)(App);
```

This also looks like the new common pattern of containers not handling any layout

File: src/store/products.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import { makeActionCreator } from '../utils/actions';
+
+const ADD = 'add product';
+const REMOVE = 'remove product';
```
Comment (HHogg) [inline src/store/products.js]: Are we not going with the standard `MODULE_ACTION_THING`?

File: src/store/products.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+export const addProduct = makeActionCreator(ADD, 'count');
+export const removeProduct = makeActionCreator(REMOVE, 'count');
```
Comment (HHogg) [inline src/store/products.js]: Much love for this new action creator 👍

File: src/index.js (hunk 2 above)
Comment (thatguynamedandy) [reply src/index.js]: Not strictly speaking, but the minute another route is added you would need it (or similar)

File: src/store/products.js (hunk 4 above)
Comment (thatguynamedandy) [reply src/store/products.js]: yes I think we should, I've started that on the redux-saga implementation, any chance we can overlook it until that goes in?

**Pull request merged**